294-documenting-assignment-process (#623)
openmsupply/application-manager-web-app: documentation/internal/Assignment-Process.md

# Assignment Process

> The logic of assigning a review spans between both web app and server, front end is only concerned with queries and mutations to review_assignment and review_question_assignment tables, whereas back end does most of the grunt work (through actions and triggers) and generates records in those tables based on user permissions and states of application and it's review.

Reviewer: ```suggestion
The logic of assigning a review spans between both web app and server — front end is only concerned with queries and mutations to review_assignment and review_question_assignment tables, whereas back end does most of the grunt work (through actions and triggers) and generates records in those tables based on user permissions and states of application and its review.
```

## Schema

Seems like a good starting (anchor) point for describing this functionality, let's have look:

![Review Assignment](images/Review-Assignment.png)

First of all, some bulletpoint (and rules) to help clarify the diagram

> - One review assignment should exists per reviewer/application/stage/level

Reviewer: ```suggestion
- One review assignment should exist per reviewer/application/stage/level
- ```

Author: f74632106a6a1af38d301b000f3fb8027bf7178e
did a manual commit, i saw an extra line in the suggested change

- Review assignment can have the following statuses
  - `Available` -> Can be assigned by assigner
  - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)

>   - `Assigned` -> At least on review_question_assignment was linked to review_assignment

Reviewer: ```suggestion
  - `Assigned` -> At least one review_question_assignment was linked to review_assignment
```

  - `Available for self-assignment` -> Can be self assigned

> - Review assignments can be restricted by section (template_section_restriction, which is an array of section codes)

Reviewer: Maybe clarify here that "restrictions" means the sections that you *can* do, but if there's not restrictions specified, they can do *all* sections.

Author: 5d80e7de004763c37418086f20a9e44f0ee99f87

> - Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the join table

Reviewer: ```suggestion
- Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the `review_assignment_assigner_join` table
```
Maybe?

> - `Assigned` status is used to indicate to the reviewer that they can start the review (if review hasn't been started already)

Reviewer: ```suggestion
- `Assigned` status indicates that the reviewer is able to create a review (or one already exists)
```

> - Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignmetn is different to application stage

Reviewer: ```suggestion
- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
```

- `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)

> ## Difference between self assignemnt and assignment

Reviewer: ```suggestion
## Difference between self-assignment and assignment
```

> - One review assignment should exist per reviewer/application/stage/level
> - Review assignment can have the following statuses
>   - `Available` -> Can be assigned by assigner
>   - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)
>   - `Assigned` -> At least one review_question_assignment was linked to review_assignment
>   - `Available for self-assignment` -> Can be self assigned
> - Review assignments can be restricted by section/s (template_section_restriction, which is an array of section codes user can review, value can be blank to allow review of all sections)
> - Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the `review_assignment_assigner_join` table
> - `Assigned` status indicates that the reviewer is able to create a review (or one already exists)
> - Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
> - `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)

Reviewer 2: Suggestion to use sub-headers to break-down get the subject of each one of this important points/rules...

```suggestion
### Assignment rules
- One `review_assignment` should exist per reviewer/application/stage/level
- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
- `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)

### Assignment status
Review assignment can have the following statuses
 - `Available` -> Can be assigned by assigner
 - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)
 - `Assigned` -> indicates that the reviewer is able to create a review (or one already exists)
   - At least one `review_question_assignment` was linked to review_assignment
 - `Available for self-assignment` -> Can be self assigned

### Restrictions
Review assignments can be restricted by section/s
 - `template_section_restriction` is an array of section codes user can review
 - Default is `null` which allow review of all sections or all questions in the application
 
### Assigner settings
- Review assignments of status `Available` or `Assigned` can be assigned by an assigner
- Assigner is linked through the `review_assignment_assigner_join` table

```

Reviewer 2: Have added sub-header and moved things up and down a little - see if that's helpful. Also I couldn't figure out from this description 2 things:

1. What is the conclusion of this sentence: "Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage."
2. "Assigner is linked through the `review_assignment_assigner_join` table" How does this work? I was trying to understand this, that's why I'm asking here also.

> The idea behind self assignment originated with requirement for users (who are not themself assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions

Reviewer: ```suggestion
The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions.
```

> ## Difference between self-assignment and assignment
> 
> The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions.

Reviewer 2: Following up what is suggested previously

```suggestion
#### Difference between self-assignment and assignment

The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. 

**Note**: Currently self assignment is not restricted by `template_section_restrictions`.
```

## The Act of Assigning - Web App

> From front end perspective, this involves creating a mutation, which adds review_question_assignments to review_assignment, changes status of review_assignment to `Assigned`. For self assignemnt trigger is set to `ON_REVIEW_SELF_ASSIGN`.

Reviewer: ```suggestion
From front end perspective, this involves creating a mutation, which adds review_question_assignments to review_assignment, and changes status of review_assignment to `Assigned`. For self assignment, trigger is set to `ON_REVIEW_SELF_ASSIGN`.
```

Front end doesn't generate review_assignment records, this happens on the server.

## Server Side

To reduce front end logic, server is responsible for generating review_assignment records, based on user permissions. This is done by an action `generateReviewAssignments` which is configured to fire when application or review changes status.

> Basically, when application is submitted, we want to look at all of the permissions linked to application template of type Assign, where template permission configurations match current level and stage, review_assignment records are generated for users who are linked to that template permission (through permission name), and any extra configurations (template section restrictions, canSelfAssign, etc..) are populated for new review_assignment record.

Reviewer: I don't think this is quite correct -- we look for permissions of type "Review" to create the review_assignment records. We look at "Assign" permissions to create review_assignment_assigner_join records.

Author: 0cdebd150db218cf5ab00558507ae5617e29b0a3
oops, thanks for noting

> ## The Act of Assigning - Web App
> 
> From front end perspective, this involves creating a mutation, which adds review_question_assignments to review_assignment, and changes status of review_assignment to `Assigned`. For self assignment, trigger is set to `ON_REVIEW_SELF_ASSIGN`.
> 
> Front end doesn't generate review_assignment records, this happens on the server.
> 
> ## Server Side
> 
> To reduce front end logic, server is responsible for generating review_assignment records, based on user permissions. This is done by an action `generateReviewAssignments` which is configured to fire when application or review changes status.
> 
> Basically, when application is submitted, we want to look at all of the permissions linked to application template of type Review, where template permission configurations match current level and stage, review_assignment records are generated for users who are linked to that template permission (through permission name), and any extra configurations (template section restrictions, canSelfAssign, etc..) are populated for new review_assignment record.
> 
> When reviews are submitted, we do similar thing but for level + 1. Typical scenario is a consolidator being informed that they can self assign a consolidation when lower level reviewer submits a review.

Reviewer 2: Moved things a lot on this suggestion. Basically deals first with what was in "Server side" with title "Pre-assignment process" and list what happens on the front end under the same title "Web app" but now with sub-titles of "Self-assignment" and "Assigning to others".

```suggestion

## Pre-assignment process

To reduce front-end logic, server is responsible for generating `review_assignment` records, based on user permissions. This is done by an action called `generateReviewAssignments` which is configured to fire when application or review changes status.
e.g.: Application is Submitted, Review is approved and moves to new Stage.

### Application submission
The server logic looks at all of the permissions linked to application template of type `review`, where template permission configurations match current `level` and `stage`. It creates `review_assignment` records for users who are linked to that template permission (through permission name), and any extra configurations (`template_section_restrictions`, `canSelfAssign`, etc..) are populated for new review_assignment record.

### Review submission
The server logic does similar thing as what is done when Application is submitted, but for level + 1. Typical scenario is a consolidator being informed that they can self assign a consolidation when lower level reviewer submits a review.

## Assigning process
Available for either a Reviewer that can do **self-assignment** or an Assigner with permission to **assign** applications of a certain template type on a `level` and `stage` (settings defined on the `template_permission` of each Template).

 To **assign** or **self-assignment** the Reviewer of sections of one application there are option available on the Applications list of on the Application home page.  

### Web App

#### Self-assignment
A GraphQL mutation is created the front-end when the user clicks on the **Self-assign** for one section.
What happens as a result of this mutation is: 
1. `review_assignment.trigger` is set to ON_REVIEW_SELF_ASSIGN.
2. Changes `review_assignment.status` to **Assigned**. 
3. Adds `review_question_assignment` (for each question of section assigned) to `review_assignment`
4. Other updates done on the server-side:
  - Update other `review_assignment.status` from **Available for self-assignment** to **Self-assigned by another**
  - Assign other sections to the same reviewer - to avoid misleading half-review.

#### Assigning to others
A GraphQL mutation  is created the front-end when the user selects one Reviewer on the drop-down list to **assign** for one section. What happens as a result of this mutation is: 
1. `review_assignment.trigger` is set to ON_REVIEW_ASSIGN.
2. Changes status of `review_assignment` to **Assigned**.
3. Adds `review_question_assignment` (for each question on the section assigned) to `review_assignment`
4. Other updates done on the server-side:
  - Update other `review_assignment.status` from **Available** to **Locked**
```

## Diagram

> This diagram is somewhat complex, it describes some functionality that is not implemented (re-assignment), it can be helpful when visualising full assignment flow (and for analysing different scenariors of assignment)

Reviewer: ```suggestion
This diagram is somewhat complex — it describes some functionality that is not implemented (re-assignment), but it can be helpful when visualising full assignment flow (and for analysing different scenarios of assignment)
```

![Assignment Flow](images/Assignment-Flow-Detailed.png)

## Re-Assignment

todo

# Front End Implementation

Best described in a diagram (this diagram is abstracted to only show bits related to assignments)

> ![Assignment UI Flow](images/Assignment-Flow-Detailed.png)

Reviewer: Links to the same image as the previous -- I think you probably meant "Assignment-UI-Flow.png"?

Author: 17356cda07839e3fb12cb9b10035e275b16d6836
